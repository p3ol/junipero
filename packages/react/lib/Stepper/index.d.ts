import { ReactNode, ComponentPropsWithRef, MutableRefObject } from 'react';

import { StepObject } from '../Step';

export declare type StepperRef = {
  active: number;
  steps: Array<StepObject>;
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface StepperProps extends ComponentPropsWithRef<any> {
  active?: number;
  steps?: Array<StepObject>;
  icon: ReactNode | JSX.Element;
  ref?: MutableRefObject<StepperRef | undefined>;
}

declare function Stepper(props: StepperProps): ReactNode | JSX.Element;

export default Stepper;
