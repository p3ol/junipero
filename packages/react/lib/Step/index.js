import { forwardRef, useImperativeHandle, useRef } from 'react';
import { classNames } from '@junipero/react';
import PropTypes from 'prop-types';

import { Check } from '../icons';

const Step = forwardRef(({ title, icon, children, status, ...rest }, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    isJunipero: true,
    innerRef,
  }));

  return (
    <div ref={innerRef} className={classNames('step', status)} {...rest}>
      <div className="content">
        <div className="step-icon">
          { status === 'completed' && (icon || <Check />) }
        </div>
        <div className="junipero extra step-details">
          <span>{ title }</span>
          <span className="junipero secondary name">{ children }</span>
        </div>
      </div>
    </div>
  );
});

Step.displayName = 'Step';
Step.propTypes = {
  title: PropTypes.string,
  status: PropTypes.string,
  icon: PropTypes.node,
};

export default Step;
