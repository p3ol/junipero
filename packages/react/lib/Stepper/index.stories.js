import Stepper from './index';

export default { title: 'react/Stepper' };

const steps = [
  { title: 'Step 1', description: 'Description 1' },
  { title: 'Step 2', description: 'Description 2' },
  { title: 'Step 3', description: 'Description 3' },
];

export const basic = () => (
  <Stepper
    steps={steps}
    currentStep={1}
  />
);

export const withIcon = () => (
  <Stepper
    steps={steps}
    currentStep={1}
    icon={<i className="junipero-icons">check_circle</i>}
  />
);
