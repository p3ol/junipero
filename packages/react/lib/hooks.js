import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
