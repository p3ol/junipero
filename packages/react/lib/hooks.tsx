import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
