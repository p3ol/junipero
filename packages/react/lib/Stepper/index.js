import {
  cloneElement,
  forwardRef,
  useMemo,
  useImperativeHandle,
  useRef,
} from 'react';
import PropTypes from 'prop-types';

import Step from '../Step';

const Stepper = forwardRef(({
  active,
  steps,
  children,
  icon,
  ...rest
}, ref) => {

  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    steps,
    active,
    icon,
    isJunipero: true,
    innerRef,
  }));

  const getStepStatus = i => {
    switch (true) {
      case i < active:
        return 'completed';
      case i === active:
        return 'active';
      default:
        return '';
    }
  };

  const availableSteps = useMemo(() => (
    steps
      ? steps.map((t, i) => (
        <Step
          status={getStepStatus(i)}
          key={i}
          title={t.title}
          icon={t.icon || icon}
        >
          { t.content }
        </Step>
      )) : children.map((t, i) =>
        cloneElement(t, { status: getStepStatus(i), key: i }
        ))
  ), [steps, children]);

  return (
    <div {...rest} ref={innerRef}>
      <div className="junipero stepper">
        { availableSteps }
      </div>
    </div>
  );
});

Stepper.displayName = 'Stepper';
Stepper.propTypes = {
  active: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  steps: PropTypes.arrayOf(PropTypes.object),
  icon: PropTypes.node,
};

export default Stepper;
