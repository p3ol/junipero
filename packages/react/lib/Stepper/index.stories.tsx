import Stepper from './index';
import Step from '../Step';

export default { title: 'react/Stepper' };

const steps = [
  { title: 'Step 1', content: 'Description 1' },
  { title: 'Step 2', content: 'Description 2' },
  { title: 'Step 3', content: 'Description 3' },
];

const stepsWithIcons = [
  {
    title: 'Step 1',
    content: 'Description 1',
    icon: <i className="junipero-icons">add_user</i>,
  },
  {
    title: 'Step 2',
    content: 'Description 2',
    icon: <i className="junipero-icons">user_edit</i>,
  },
  {
    title: 'Step 3',
    content: 'Description 3',
    icon: <i className="junipero-icons">user_valid</i>,
  },
];

export const Basic = () => (
  <>
    <Stepper active={0}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={1}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={2}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={3}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
  </>
);

export const WithStepsProp = () => (
  <>
    <Stepper
      steps={steps}
      active={0}
    />
    <Stepper
      steps={steps}
      active={1}
    />
    <Stepper
      steps={steps}
      active={2}
    />
    <Stepper
      steps={steps}
      active={3}
    />
  </>
);

export const WithDefaultIcon = () => (
  <>
    <Stepper
      steps={steps}
      active={0}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={1}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={2}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={3}
      icon={<i className="junipero-icons">check_circle</i>}
    />
  </>
);

export const WithCustomIcons = () => (
  <>
    <Stepper
      steps={stepsWithIcons}
      active={0}
    />
    <Stepper
      steps={stepsWithIcons}
      active={1}
    />
    <Stepper
      steps={stepsWithIcons}
      active={2}
    />
    <Stepper
      steps={stepsWithIcons}
      active={3}
    />
  </>
);
