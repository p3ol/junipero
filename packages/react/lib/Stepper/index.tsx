import {
  type RefObject,
  type ReactNode,
  Children,
  cloneElement,
  useMemo,
  useImperativeHandle,
  useRef,
  use,
} from 'react';

import type {
  JuniperoRef,
  ReactElt,
  ReactLazy,
  SpecialComponentPropsWithRef
} from '../types';
import Step, { type StepObject } from '../Step';

export declare interface StepperRef extends JuniperoRef {
  active: number;
  steps: StepObject[];
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface StepperProps
  extends SpecialComponentPropsWithRef<'div', StepperRef> {
  active?: number;
  steps?: StepObject[];
  icon?: ReactNode;
}

const Stepper = ({
  ref,
  active,
  steps,
  children,
  icon,
  ...rest
}: StepperProps) => {
  const innerRef = useRef<HTMLDivElement>(null);

  useImperativeHandle(ref, () => ({
    steps,
    active,
    icon,
    isJunipero: true,
    innerRef,
  }));

  const getStepStatus = (i: number) => {
    switch (true) {
      case i < active:
        return 'completed';
      case i === active:
        return 'active';
      default:
        return null;
    }
  };

  const availableSteps = useMemo(() => (
    steps ? steps.map((t, i) => (
      <Step
        status={getStepStatus(i)}
        key={i}
        title={t.title}
        icon={t.icon || icon}
      >
        { t.content }
      </Step>
    )) : Children.toArray(children).map((
      t: ReactElt,
      i: number
    ) => (
      cloneElement(
        (t as unknown as ReactLazy ).$$typeof === Symbol.for('react.lazy')
          ? use<ReactElt>((t as unknown as ReactLazy)._payload) : t as ReactElt,
        { status: getStepStatus(i), key: i })
     ))
  ), [steps, children]);

  return (
    <div { ...rest } ref={innerRef}>
      <div className="junipero stepper">
        { availableSteps }
      </div>
    </div>
  );
};

Stepper.displayName = 'Stepper';

export default Stepper;
