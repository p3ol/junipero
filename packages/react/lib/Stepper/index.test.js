import { render } from '@testing-library/react';

import Stepper from './';
import Step from '../Step';

describe('<Stepper />', () => {

  const steps = [
    { title: 'Step 1', description: 'Description 1' },
    { title: 'Step 2', description: 'Description 2' },
    { title: 'Step 3', description: 'Description 3' },
  ];

  const stepsWithIcons = [
    {
      title: 'Step 1',
      content: 'Description 1',
      icon: <i className="junipero-icons">add_user</i>,
    },
    {
      title: 'Step 2',
      content: 'Description 2',
      icon: <i className="junipero-icons">user_edit</i>,
    },
    {
      title: 'Step 3',
      content: 'Description 3',
      icon: <i className="junipero-icons">user_valid</i>,
    },
  ];

  it('should render steps prop with each status', () => {
    const { unmount, container } = render(
      <Stepper steps={steps} active={1} />
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should render step children with each status', () => {
    const { unmount, container } = render(
      <Stepper active={1}>
        <Step title="Step 1">Description 1</Step>
        <Step title="Step 2">Description 2</Step>
        <Step title="Step 3">Description 3</Step>
      </Stepper>
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should display a different icon per step', () => {
    const { unmount, container } = render(
      <Stepper
        steps={stepsWithIcons}
        active={0}
        icon={<i className="junipero-icons">check_circle</i>}
      />
    );
    expect(container).toMatchSnapshot();
    unmount();
  });
});
