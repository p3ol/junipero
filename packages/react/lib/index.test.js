import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
