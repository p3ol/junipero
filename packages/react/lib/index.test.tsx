import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
