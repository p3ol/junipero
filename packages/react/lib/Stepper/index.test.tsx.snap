// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Stepper /> should display a different icon per step 1`] = `
<div>
  <div>
    <div
      class="junipero stepper"
    >
      <div
        class="step active"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 1
            </span>
            <span
              class="junipero secondary name"
            >
              Description 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="step"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 2
            </span>
            <span
              class="junipero secondary name"
            >
              Description 2
            </span>
          </div>
        </div>
      </div>
      <div
        class="step"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 3
            </span>
            <span
              class="junipero secondary name"
            >
              Description 3
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Stepper /> should render step children with each status 1`] = `
<div>
  <div>
    <div
      class="junipero stepper"
    >
      <div
        class="step completed"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          >
            <svg
              class="junipero icon check"
              height="7"
              viewBox="0 0 10 7"
              width="10"
            >
              <path
                d="M1.75259 3.49582L4.25675 6L8.24741 1"
              />
            </svg>
          </div>
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 1
            </span>
            <span
              class="junipero secondary name"
            >
              Description 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="step active"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 2
            </span>
            <span
              class="junipero secondary name"
            >
              Description 2
            </span>
          </div>
        </div>
      </div>
      <div
        class="step"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 3
            </span>
            <span
              class="junipero secondary name"
            >
              Description 3
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Stepper /> should render steps prop with each status 1`] = `
<div>
  <div>
    <div
      class="junipero stepper"
    >
      <div
        class="step completed"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          >
            <svg
              class="junipero icon check"
              height="7"
              viewBox="0 0 10 7"
              width="10"
            >
              <path
                d="M1.75259 3.49582L4.25675 6L8.24741 1"
              />
            </svg>
          </div>
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 1
            </span>
            <span
              class="junipero secondary name"
            >
              Description 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="step active"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 2
            </span>
            <span
              class="junipero secondary name"
            >
              Description 2
            </span>
          </div>
        </div>
      </div>
      <div
        class="step"
      >
        <div
          class="content"
        >
          <div
            class="step-icon"
          />
          <div
            class="junipero extra step-details"
          >
            <span>
              Step 3
            </span>
            <span
              class="junipero secondary name"
            >
              Description 3
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
